cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java

```diff
@@ -170,22 +219,96 @@ private void ensureTopicsCreated(Map<String, ?> config) {
+      String rack = bm.rack().isEmpty() ? String.valueOf(bm.id()) : bm.rack().get();
+      brokersByRack.putIfAbsent(rack, new ArrayList<>());
+      brokersByRack.get(rack).add(bm.id());
+      rackByBroker.put(bm.id(), rack);
+    }
+    if (replicationFactor > brokersByRack.size()) {
+      throw new RuntimeException("Unable to increase topic " + topic + " replica factor to " + replicationFactor
+                                 + " since there are only " + brokersByRack.size() + " racks in the cluster.");
+    }
+
+    scala.collection.mutable.Map<Object, Seq<Object>> newReplicaAssignment = new scala.collection.mutable.HashMap<>();
+    List<String> racks = new ArrayList<>(brokersByRack.keySet());
+    int [] cursors = new int[racks.size()];
+    int rackCursor = 0;
+    for (MetadataResponse.PartitionMetadata pm : topicMetadata.partitionMetadata()) {
+      List<Object> newAssignedReplica = new ArrayList<>();
+      Set<String> currentOccupiedRack = new HashSet<>();
+      // Make sure the current replicas are in new replica list.
+      pm.replicas().forEach(node -> {
+        newAssignedReplica.add(node.id());
+        currentOccupiedRack.add(rackByBroker.get(node.id()));
+      });
+      // Add new replica to partition in rack-aware(if rack info is available), round-robin way.
+      while (newAssignedReplica.size() < replicationFactor) {
+        if (!currentOccupiedRack.contains(racks.get(rackCursor))) {
+          String rack = racks.get(rackCursor);
+          int cursor = cursors[rackCursor];
+          newAssignedReplica.add(brokersByRack.get(rack).get(cursor));
+          cursors[rackCursor] = (cursor + 1) % brokersByRack.get(rack).size();
+        }
+        rackCursor = (rackCursor + 1) % racks.size();
+      }
+      newReplicaAssignment.put(pm.partition(), JavaConverters.asScalaIteratorConverter(newAssignedReplica.iterator()).asScala().toSeq());
+    }
+    AdminUtils.createOrUpdateTopicPartitionAssignmentPathInZK(zkUtils, topic, newReplicaAssignment, props, true);
+    LOG.info("The replication factor of Kafka topic " + topic + " has increased to " + replicationFactor + ".");
+  }
+
+  private void ensureTopicCreated(ZkUtils zkUtils, Set<String> allTopics, String topic, long retentionMs, int replicationFactor, int partitionCount) {
     Properties props = new Properties();
     props.setProperty(LogConfig.RetentionMsProp(), Long.toString(retentionMs));
     props.setProperty(LogConfig.CleanupPolicyProp(), DEFAULT_CLEANUP_POLICY);
     if (!allTopics.contains(topic)) {
-      AdminUtils.createTopic(zkUtils, topic, 32, replicationFactor, props, RackAwareMode.Safe$.MODULE$);
+      AdminUtils.createTopic(zkUtils, topic, partitionCount, replicationFactor, props, RackAwareMode.Safe$.MODULE$);
     } else {
+      ensureTopicNotUnderPartitionReassignment(zkUtils, topic);
       AdminUtils.changeTopicConfig(zkUtils, topic, props);
+      MetadataResponse.TopicMetadata topicMetadata = AdminUtils.fetchTopicMetadataFromZk(
```

Comment:
Why don't we use `MetadataResponse.TopicMetadata topicMetadata = AdminUtils.fetchTopicMetadataFromZk(topic, zkUtils);` here?

Reply:
This is because in a later version of the Kafka,  method `fetchTopicMetadataFromZk(topic: String, zkUtils: ZkUtils)` is removed from AdminUtils. So to make CC compatible with more Kafka versions, I switch to this method(which is available in more versions).